File: web/src/game/player.ts
// src/game/player.ts
// Player: move/aim/shoot/XP/upgrade multipliers. Mirrors player.py.

import {
  PLAYER_BASE_MAX_HP,
  PLAYER_BASE_SPEED,
  PLAYER_RADIUS,
  WORLD_HEIGHT,
  WORLD_WIDTH,
  XP_BASE_REQUIREMENT,
} from "./settings";
import { moveCircle } from "./collision";
import { clamp } from "./utils";
import { WeaponManager } from "./weapon";
import type { WeaponData } from "./data";
import { Bullet } from "./bullet";
import type { IGame } from "./types";
import type { Vec } from "./vec";
import type { Camera } from "./camera";
import { drawPlayerSprite } from "./pixelArt";

export interface PlayerOpts {
  unlocked?: string[];
  coins?: number;
  level?: number;
  xp?: number;
  weaponData: Record<string, WeaponData>;
  weaponMods?: Record<string, string[]>;
  previewOnly?: boolean;
  username?: string;
}

export class Player {
  radius = PLAYER_RADIUS;
  static XP_BASE = XP_BASE_REQUIREMENT;

  pos: Vec;
  vel: Vec = { x: 0, y: 0 };
  angle = 0;
  username?: string;

  maxHp = PLAYER_BASE_MAX_HP;
  hp = PLAYER_BASE_MAX_HP;
  armor = 0;
  baseSpeed = PLAYER_BASE_SPEED;
  coins = 0;

  level = 1;
  xp = 0;
  pendingLevels = 0;
  upgradeLevels: Record<string, number> = {};

  damageMult = 1;
  fireRateMult = 1;
  reloadMult = 1;
  speedMult = 1;
  critBonus = 0;
  critMultBonus = 0;
  regen = 0;
  magnetMult = 1;
  lifeSteal = 0;
  pierceBonus = 0;
  skillPoints = 0;
  /** Set by a tank's charge attack: blocks movement/firing while > 0. */
  stunTimer = 0;

  // Drone companion ("UFO") — unlocked by buying it in the shop.
  hasDrone = false;
  droneAngle = 0;
  droneCooldown = 0;
  droneDamage = 18;

  weapons: WeaponManager;

  flashTimer = 0;
  recoilTimer = 0;
  emptyClickTimer = 0;
  invuln = 0;
  walkCycle = 0;
  moving = false;
  dustCd = 0;
  dead = false;
  previewOnly = false;

  constructor(pos: Vec, opts: PlayerOpts) {
    this.pos = { ...pos };
    this.coins = Math.max(0, Math.floor(opts.coins ?? 0));
    this.level = Math.max(1, Math.floor(opts.level ?? 1));
    this.xp = Math.max(0, Math.floor(opts.xp ?? 0));
    this.weapons = new WeaponManager(opts.unlocked ?? ["pistol"], opts.weaponData, opts.weaponMods ?? {});
    this.previewOnly = !!opts.previewOnly;
    this.username = opts.username;
  }

  get xpNeeded(): number {
    return Player.XP_BASE * this.level;
  }

  get speed(): number {
    return this.baseSpeed * this.speedMult;
  }

  heal(amount: number): void {
    this.hp = Math.min(this.maxHp, this.hp + amount);
  }

  addArmor(amount: number): void {
    this.armor = Math.min(100, this.armor + amount);
  }

  /** Push the player away from `source`, sliding along walls like a zombie's knockback. */
  knockbackFrom(source: Vec, force: number, game: IGame): void {
    const dx = this.pos.x - source.x;
    const dy = this.pos.y - source.y;
    const d = Math.hypot(dx, dy) || 1;
    const rects = game.map!.getNear(this.pos, this.radius + 4);
    moveCircle(this.pos, { x: (dx / d) * force, y: (dy / d) * force }, this.radius, rects);
    this.pos.x = clamp(this.pos.x, this.radius, WORLD_WIDTH - this.radius);
    this.pos.y = clamp(this.pos.y, this.radius, WORLD_HEIGHT - this.radius);
  }

  update(dt: number, game: IGame): void {
    const inp = game.input;
    this.stunTimer = Math.max(0, this.stunTimer - dt);
    const stunned = this.stunTimer > 0;
    const move: Vec = stunned
      ? { x: 0, y: 0 }
      : {
          x: (inp.isDown("right") ? 1 : 0) - (inp.isDown("left") ? 1 : 0),
          y: (inp.isDown("down") ? 1 : 0) - (inp.isDown("up") ? 1 : 0),
        };
    this.moving = !stunned && (move.x !== 0 || move.y !== 0);
    if (this.moving) {
      const len = Math.hypot(move.x, move.y) || 1;
      move.x /= len;
      move.y /= len;
      const rects = game.map!.getNear(this.pos, this.radius + 4);
      moveCircle(
        this.pos,
        { x: move.x * this.speed * dt, y: move.y * this.speed * dt },
        this.radius,
        rects,
      );
      this.pos.x = clamp(this.pos.x, this.radius, WORLD_WIDTH - this.radius);
      this.pos.y = clamp(this.pos.y, this.radius, WORLD_HEIGHT - this.radius);
      const lastCycle = this.walkCycle;
      this.walkCycle += dt * 10;
      
      const stepFreq = Math.PI;
      if (Math.floor(lastCycle / stepFreq) !== Math.floor(this.walkCycle / stepFreq)) {
        const stepSound = Math.random() < 0.5 ? "player.footstep1" : "player.footstep2";
        game.audio.playSFX(stepSound, this.pos);
      }
      // FOOTSTEP DUST (optional, OFF by default): subtle puffs trailing the
      // feet every 0.18s while moving — only when the setting is enabled.
      if (game.save.settings.footstep_dust) {
        this.dustCd -= dt;
        if (this.dustCd <= 0) {
          this.dustCd = 0.18;
          game.particles.dust({
            x: this.pos.x + Math.cos(this.walkCycle + Math.PI) * 8,
            y: this.pos.y + Math.sin(this.walkCycle + Math.PI) * 8,
          });
        }
      }
    }

    // Aim: touch auto-aim override (mobile) or mouse in world coords (desktop).
    const aimWorld = inp.getAimWorld(game.camera);
    this.angle = Math.atan2(aimWorld.y - this.pos.y, aimWorld.x - this.pos.x);

    this.emptyClickTimer = Math.max(0, this.emptyClickTimer - dt);

    // Weapon switching.
    const prevWepId = this.weapons.currentId;
    if (inp.isPressed("weapon1")) this.weapons.selectSlot(1);
    if (inp.isPressed("weapon2")) this.weapons.selectSlot(2);
    if (inp.isPressed("weapon3")) this.weapons.selectSlot(3);
    if (inp.isPressed("weapon4")) this.weapons.selectSlot(4);
    if (inp.isPressed("weapon5")) this.weapons.selectSlot(5);
    if (inp.isPressed("weapon6")) this.weapons.selectSlot(6);
    if (inp.isPressed("weapon7")) this.weapons.selectSlot(7);
    if (inp.isPressed("weapon8")) this.weapons.selectSlot(8);
    if (inp.isPressed("next_weapon")) this.weapons.cycle();

    if (this.weapons.currentId !== prevWepId) {
      game.audio.cancelReloadSound(prevWepId);
      const prevWep = this.weapons.weapons[prevWepId];
      if (prevWep && prevWep.reloading) {
        prevWep.reloading = false;
      }
      game.audio.playSFX("ui.equip");
    }

    const w = this.weapons.current;
    const wantFire = inp.mouseHeld;
    if (
      inp.isPressed("reload") ||
      (wantFire && !w.reloading && w.ammo === 0 && w.reserve > 0)
    ) {
      if (w.startReload(this.reloadMult)) {
        game.audio.playSFX(`weapon.${w.id}.reload`, this.pos);
      }
    }
    if (!stunned && wantFire && w.canFire(wantFire || w.auto)) {
      this.fire(game);
    } else if (!stunned && wantFire && !w.reloading && w.ammo === 0 && w.cooldown <= 0 && this.emptyClickTimer <= 0) {
      game.audio.playSFX(`weapon.${w.id}.empty`, this.pos);
      this.emptyClickTimer = 0.18;
    }
    this.weapons.update(dt);

    this.flashTimer = Math.max(0, this.flashTimer - dt);
    this.recoilTimer = Math.max(0, this.recoilTimer - dt);
    this.invuln = Math.max(0, this.invuln - dt);
    if (this.regen > 0) this.heal(this.regen * dt);
    if (this.hasDrone) this.droneTick(dt, game);
  }

  /** Orbiting drone that auto-fires at the nearest zombie in range. */
  private droneTick(dt: number, game: IGame): void {
    this.droneAngle += dt * 1.6;
    this.droneCooldown -= dt;
    if (this.droneCooldown > 0) return;
    for (const z of game.zombies) {
      if (z.dying) continue;
      const dx = z.pos.x - this.pos.x;
      const dy = z.pos.y - this.pos.y;
      if (dx * dx + dy * dy < 280 * 280) {
        const muzzle: Vec = {
          x: this.pos.x + Math.cos(this.droneAngle + Math.PI) * 36,
          y: this.pos.y + Math.sin(this.droneAngle + Math.PI) * 36,
        };
        const ang = Math.atan2(z.pos.y - muzzle.y, z.pos.x - muzzle.x);
        game.bullets.push(
          new Bullet(muzzle, ang, 1200, this.droneDamage, "player"),
        );
        game.particles.muzzleFlash(muzzle, ang);
        this.droneCooldown = 0.55;
        return;
      }
    }
  }

  private fire(game: IGame): void {
    const w = this.weapons.current;
    const muzzle: Vec = {
      x: this.pos.x + Math.cos(this.angle) * 24,
      y: this.pos.y + Math.sin(this.angle) * 24,
    };
    const shots = w.fire(
      this.angle,
      this.damageMult,
      this.critBonus,
      this.critMultBonus,
    );
    w.cooldown = w.fireRate / Math.max(0.01, this.fireRateMult);
    for (const s of shots) {
      const b = new Bullet(
        muzzle,
        s.angle,
        s.speed,
        s.damage,
        "player",
        s.crit,
        s.radius,
        s.lifetime,
        s.elem,
      );
      if (s.elem === "pierce") {
        // Crossbow bolt: pierces 3 enemies by default, +1 per pierce skill.
        b.pierceLeft = 3 + this.pierceBonus;
      }
      game.bullets.push(b);
    }
    game.particles.muzzleFlash(muzzle, this.angle);
    this.recoilTimer = 0.09;
    game.camera.shake(w.pellets === 1 ? 1.5 : 4);
    const sbw = game.stats.shots_by_weapon;
    sbw[w.id] = (sbw[w.id] ?? 0) + 1;
    game.stats.shots_fired = (game.stats.shots_fired ?? 0) + shots.length;
    game.audio.playSFX(`weapon.${w.id}.shoot`, this.pos);
  }

  takeDamage(amount: number, game: IGame): void {
    if (this.invuln > 0 || this.dead) return;
    const absorbed = Math.min(this.armor, amount);
    this.armor -= absorbed;
    const hpLoss = amount - absorbed;
    this.hp -= hpLoss;
    this.flashTimer = game.save.settings.hit_effects ? 0.25 : 0;
    this.invuln = 0.15;
    game.particles.blood(this.pos, 6, undefined, game.save.settings.hit_effects);
    game.particles.damageNumber(this.pos, hpLoss + absorbed, false, game.save.settings.damage_numbers);
    game.camera.shake(Math.min(10, 2 + amount * 0.15));
    game.audio.playSFX("player.damage", this.pos);
    if (this.hp <= 0) {
      this.hp = 0;
      this.dead = true;
    }
  }    addXp(amount: number, game: IGame): void {
    this.xp += amount;
    while (this.xp >= this.xpNeeded) {
      this.xp -= this.xpNeeded;
      this.level += 1;
      this.pendingLevels += 1;
      this.skillPoints += 1;
      game.onLevelUp();
    }
    game.save.data["player_level"] = this.level;
    game.save.data["xp"] = this.xp;
  }

  draw(ctx: CanvasRenderingContext2D, cam: Camera): void {
    const sp = cam.apply(this.pos);
    
    // Draw username above player if it exists
    if (this.username) {
      ctx.fillStyle = "#EBEBE1";
      ctx.font = "bold 12px ui-monospace, monospace";
      ctx.textAlign = "center";
      ctx.textBaseline = "bottom";
      ctx.fillText(this.username, sp.x, sp.y - this.radius - 10);
    }

    drawPlayerSprite(
      ctx,
      sp,
      this.angle,
      this.moving ? this.walkCycle : 0,
      this.recoilTimer,
      this.weapons.currentId,
      this.flashTimer > 0 && Math.floor(this.flashTimer * 20) % 2 === 0,
    );

    // Drone companion: orbits the player and glows softly.
    if (this.hasDrone) {
      const dxx = sp.x + Math.cos(this.droneAngle) * 40;
      const dyy = sp.y + Math.sin(this.droneAngle) * 40 - 14;
      ctx.strokeStyle = "rgba(140, 230, 255, 0.45)";
      ctx.lineWidth = 1;
      ctx.beginPath();
      ctx.arc(dxx, dyy, 11, 0, Math.PI * 2);
      ctx.stroke();
      ctx.fillStyle = "#3A3A44";
      ctx.beginPath();
      ctx.arc(dxx, dyy, 8, 0, Math.PI * 2);
      ctx.fill();
      ctx.fillStyle = "#8FE8FF";
      ctx.beginPath();
      ctx.arc(dxx, dyy, 5, 0, Math.PI * 2);
      ctx.fill();
      ctx.fillStyle = "#EAFBFF";
      ctx.beginPath();
      ctx.arc(dxx - 2, dyy - 2, 2, 0, Math.PI * 2);
      ctx.fill();
    }
  }
}

File: player.py
"""Player: movement, aiming, shooting, XP/level, upgrades, damage."""
from __future__ import annotations

import math

import pygame

import settings as S
import weapon as wp_mod
from collision import move_circle
from sprite.sprite_cache import get_player_sprite
from weapon import WeaponManager
from utils import clamp as S_clamp


class Player:
    radius = S.PLAYER_RADIUS
    XP_BASE = S.XP_BASE_REQUIREMENT

    def __init__(self, pos: pygame.Vector2, unlocked: list[str] | None = None,
                 coins: int = 0, level: int = 1, xp: int = 0) -> None:
        self.pos = pygame.Vector2(pos)
        self.vel = pygame.Vector2(0, 0)
        self.angle = 0.0

        self.max_hp = S.PLAYER_BASE_MAX_HP
        self.hp = self.max_hp
        self.armor = 0.0
        self.base_speed = S.PLAYER_BASE_SPEED
        self.coins = int(coins)

        self.level = max(1, int(level))
        self.xp = int(xp)
        self.pending_levels = 0
        self.upgrade_levels: dict[str, int] = {}

        # Upgrade multipliers (stacked by UpgradeSystem).
        self.damage_mult = 1.0
        self.fire_rate_mult = 1.0     # cooldown is divided by this
        self.reload_mult = 1.0        # reload time multiplied by this
        self.speed_mult = 1.0
        self.crit_bonus = 0.0
        self.crit_mult_bonus = 0.0
        self.regen = 0.0
        self.magnet_mult = 1.0
        self.life_steal = 0.0
        self.pierce_bonus = 0
        self.skill_points = 0

        self.weapons = WeaponManager(unlocked or ["pistol"])

        self.flash_timer = 0.0
        self.invuln = 0.0
        self.walk_cycle = 0.0
        self.moving = False
        self.dead = False
        self.drone_angle = 0.0
        self.drone_cooldown = 0.0
        self.drone_damage = 18.0

    # ------------------------------------------------------------ helpers --
    @property
    def xp_needed(self) -> int:
        return self.XP_BASE * self.level

    @property
    def speed(self) -> float:
        return self.base_speed * self.speed_mult

    def heal(self, amount: float) -> None:
        self.hp = min(self.max_hp, self.hp + amount)

    def add_armor(self, amount: float) -> None:
        self.armor = min(100.0, self.armor + amount)

    # ------------------------------------------------------------- update --
    def update(self, dt: float, game) -> None:
        inp = game.input

        move = pygame.Vector2(
            (1 if inp.is_down("move_right") else 0) -
            (1 if inp.is_down("move_left") else 0),
            (1 if inp.is_down("move_down") else 0) -
            (1 if inp.is_down("move_up") else 0),
        )
        self.moving = move.length_squared() > 0
        if self.moving:
            move.normalize()
            rects = game.map.get_near(self.pos, self.radius + 4)
            move_circle(self.pos, move * self.speed * dt, self.radius, rects)
            self.pos.x = S_clamp(self.pos.x, self.radius,
                                 S.WORLD_WIDTH - self.radius)
            self.pos.y = S_clamp(self.pos.y, self.radius,
                                 S.WORLD_HEIGHT - self.radius)
            self.walk_cycle += dt * 10

        mouse_world = cam_aim(game)
        self.angle = math.atan2(mouse_world.y - self.pos.y,
                                mouse_world.x - self.pos.x)

        # Weapon switching. Routed through a single switchWeapon() entry
        # point so reload cancellation, validation and HUD feedback all
        # run through the same code path. Only listen during gameplay;
        # when paused / leveling up the parent loop has already returned
        # before reaching Player.update(), but we also gate by game.state
        # so an in-flight key event from the upgrade menu never leaks
        # into a switch on the first frame back in PLAYING.
        if getattr(game, "state", None) == "PLAYING":
            for slot in range(1, 6):
                action = f"weapon_{slot}"
                if inp.is_pressed(action) or inp.key_pressed(
                        pygame.K_1 + slot - 1):
                    self.switchWeapon(slot)
            if inp.key_pressed(pygame.K_q):     # Q to cycle weapon
                self.weapons.cycle()

        w = self.weapons.current
        want_fire = 1 in inp.mouse_down
        if inp.is_pressed("reload") or (
                want_fire and not w.reloading and w.ammo == 0 and w.reserve > 0):
            if w.start_reload(self.reload_mult):
                game.audio.play("reload")
        if want_fire and w.can_fire(want_fire or w.auto):
            self._fire(game)
        self.weapons.update(dt)

        self.flash_timer = max(0.0, self.flash_timer - dt)
        self.invuln = max(0.0, self.invuln - dt)
        if self.regen > 0 and not self.dead:
            self.heal(self.regen * dt)
        if self.moving and not self.dead:
            # Respect the `footstep_dust` setting so players can turn off
            # the moving dust puffs if they find them distracting. The
            # setting defaults to False (off) to avoid a swarm of bright
            # dots orbiting the player every 0.18s.
            _footstep_on = bool(
                (getattr(game, "save", None) is not None)
                and game.save.settings.get("footstep_dust", False)
            )
            if _footstep_on:
                self._dust_cd = getattr(self, "_dust_cd", 0.0) - dt
                if self._dust_cd <= 0:
                    self._dust_cd = 0.18
                    offset = pygame.Vector2(
                        math.cos(self.walk_cycle + math.pi) * 8,
                        math.sin(self.walk_cycle + math.pi) * 8)
                    game.particles.dust(self.pos + offset)
        self._drone_tick(dt, game)

    def _fire(self, game) -> None:
        from bullet import Bullet
        w = self.weapons.current
        muzzle = self.pos + pygame.Vector2(math.cos(self.angle),
                                           math.sin(self.angle)) * 24
        shots = w.fire(self.angle,
                       damage_mult=self.damage_mult,
                       crit_bonus=self.crit_bonus,
                       crit_mult_bonus=self.crit_mult_bonus)
        w.cooldown = w.fire_rate / max(0.01, self.fire_rate_mult)
        for spec in shots:
            radius = 6.0 if w.id in ("flamethrower",) else 4.0
            b = Bullet(muzzle.copy(), spec["angle"],
                       spec["speed"], spec["damage"],
                       owner="player", crit=spec["crit"], radius=radius)
            if w.id == "flamethrower":
                b.elem = "fire"
                b.lifetime = 0.55
            elif w.id == "plasma":
                b.elem = "plasma"
                b.radius = 7.0
            elif w.id == "crossbow":
                b.elem = "pierce"
            game.bullets.append(b)
        if w.id == "flamethrower":
            game.particles.burn_trail(muzzle, self.angle, 8)
            game.lighting.add_light(muzzle, 120, (255, 130, 60), 0.8)
        elif w.id == "plasma":
            game.particles.muzzle_flash(muzzle, self.angle)
            game.lighting.add_light(muzzle, 90, (170, 100, 255), 0.6)
        else:
            game.particles.muzzle_flash(muzzle, self.angle)
            game.lighting.add_light(muzzle, 70, (255, 220, 130), 0.4)
        game.shake_camera(1.5 if w.pellets == 1 else 4.0)
        game.stats["shots_by_weapon"][w.id] = \
            game.stats["shots_by_weapon"].get(w.id, 0) + 1
        sound_map = {
            "shotgun": "shotgun", "sniper": "sniper",
            "smg": "smg", "rifle": "rifle",
            "flamethrower": "shoot", "plasma": "rifle",
            "crossbow": "sniper",
        }
        game.audio.play(sound_map.get(w.id, "shoot"))

    # ----------------------------------------------------------- switch ----
    def switchWeapon(self, slot: int) -> bool:
        """Public entry point for weapon switching.

        1 <= slot <= 5. Locked / out-of-range slots fall back silently
        to the current weapon so pressing a wrong key never crashes and
        never deselects the active gun. Cancel any in-progress reload
        on the previous weapon so ammo doesn't bleed into the new one.
        """
        try:
            slot_i = int(slot)
        except (TypeError, ValueError):
            return False
        if not (1 <= slot_i <= 5):
            return False
        changed = self.weapons.switch_weapon(slot_i)
        # After switch: only the new weapon's cooldown is ticked by
        # update(), so a stale cooldown on the old weapon is harmless
        # — but we still clear it so the old gun is "fresh" when the
        # player comes back to it.
        if changed:
            old_ids = [wid for wid in self.weapons.unlocked_in_order()
                       if wid != self.weapons.current_id]
            for wid in old_ids:
                wp = self.weapons.weapons.get(wid)
                if wp is not None:
                    wp.cooldown = 0.0
            w = self.weapons.current
            if w.ammo > 0:
                w.cooldown = 0.0
        return changed

    # ------------------------------------------------------------- combat --
    def take_damage(self, amount: float, game) -> None:
        if self.invuln > 0 or self.dead:
            return
        absorbed = min(self.armor, amount)
        self.armor -= absorbed
        hp_loss = amount - absorbed
        self.hp -= hp_loss
        self.flash_timer = 0.25
        self.invuln = 0.15
        game.particles.blood(self.pos, 6)
        game.particles.damage_number(self.pos, hp_loss + absorbed, False)
        game.shake_camera(min(10, 2 + amount * 0.15))
        game.audio.play("player_hit")
        if self.hp <= 0:
            self.hp = 0
            self.dead = True

    def add_xp(self, amount: int, game) -> None:
        self.xp += amount
        while self.xp >= self.xp_needed:
            self.xp -= self.xp_needed
            self.level += 1
            self.pending_levels += 1
            self.skill_points += 1
            game.on_level_up()
        game.save.data["player_level"] = self.level
        game.save.data["xp"] = self.xp

    # ----------------------------------------------------------- drone ----
    def _drone_tick(self, dt: float, game) -> None:
        self.drone_angle += dt * 1.6
        self.drone_cooldown -= dt
        if self.drone_cooldown > 0:
            return
        for z in game.zombies:
            if z.pos.distance_to(self.pos) < 280:
                from bullet import Bullet
                muzzle = self.pos + pygame.Vector2(
                    math.cos(self.drone_angle + math.pi),
                    math.sin(self.drone_angle + math.pi)) * 36
                angle = math.atan2(z.pos.y - muzzle.y,
                                   z.pos.x - muzzle.x)
                game.bullets.append(Bullet(muzzle, angle, 1200,
                                           self.drone_damage,
                                           owner="player"))
                game.particles.muzzle_flash(muzzle, angle)
                self.drone_cooldown = 0.55
                break

    # --------------------------------------------------------------- draw --
    def draw(self, surface: pygame.Surface, cam) -> None:
        sp = cam.apply(self.pos)
        bob = math.sin(self.walk_cycle) * 1.5 if self.moving else 0
        frame = int(self.walk_cycle) % 4 if self.moving else 0
        sprite = get_player_sprite(frame)
        flash = (self.flash_timer > 0
                 and int(self.flash_timer * 20) % 2 == 0)
        if flash:
            tinted = sprite.copy()
            white = pygame.Surface(tinted.get_size(), pygame.SRCALPHA)
            white.fill((255, 255, 255, 110))
            tinted.blit(white, (0, 0))
            sprite = tinted
        rotated = pygame.transform.rotate(sprite, -math.degrees(self.angle))
        rect = rotated.get_rect(center=(int(sp.x), int(sp.y) - int(bob)))
        surface.blit(rotated, rect)

        if self.invuln > 0 and int(self.invuln * 20) % 2 == 0:
            ring = pygame.Surface((self.radius * 4, self.radius * 4),
                                  pygame.SRCALPHA)
            pygame.draw.circle(ring, (180, 220, 255, 160),
                               (self.radius * 2, self.radius * 2),
                               self.radius + 4, 2)
            surface.blit(ring, ring.get_rect(center=(int(sp.x), int(sp.y))))

        if self.armor > 0:
            ax = sp.x + math.cos(self.angle) * 18
            ay = sp.y + math.sin(self.angle) * 18 - 18
            pygame.draw.circle(surface, (90, 180, 255), (int(ax), int(ay)), 6)
            pygame.draw.circle(surface, (220, 240, 255), (int(ax), int(ay)), 3)

        drone_x = sp.x + math.cos(self.drone_angle) * 38
        drone_y = sp.y + math.sin(self.drone_angle) * 38 - 16
        pygame.draw.circle(surface, (60, 60, 70), (int(drone_x), int(drone_y)),
                           9, 2)
        pygame.draw.circle(surface, (140, 220, 255), (int(drone_x), int(drone_y)),
                           6)
        pygame.draw.circle(surface, (220, 240, 255), (int(drone_x - 2),
                                                       int(drone_y - 2)), 2)


def cam_aim(game) -> pygame.Vector2:
    """Mouse position converted to world coordinates."""
    return game.camera.screen_to_world(pygame.mouse.get_pos())
